consensus/src/simplex/actors/voter/actor.rs

```diff
@@ -161,55 +182,137 @@ impl<E: Clock, S: Scheme, D: Digest> Round<E, S, D> {
     pub fn set_leader(&mut self, seed: Option<S::Seed>) {
         let (leader, leader_idx) =
             select_leader::<S, _>(self.scheme.participants().as_ref(), self.round, seed);
-        self.leader = Some(leader_idx);
-
         debug!(round=?self.round, ?leader, ?leader_idx, "leader elected");
+
+        self.leader = Some(Leader {
+            idx: leader_idx,
+            key: leader,
+        });
     }
 
-    fn add_recovered_proposal(&mut self, proposal: Proposal<D>) {
-        if self.proposal.is_none() {
-            debug!(?proposal, "setting verified proposal");
-            self.proposal = Some(proposal);
-        } else if let Some(previous) = &self.proposal {
-            assert_eq!(proposal, *previous);
+    /// Returns the equivocator if the new proposal overrides an existing one.
+    fn add_recovered_proposal(&mut self, proposal: Proposal<D>) -> Option<S::PublicKey> {
+        // Check for equivocation
+        if self.proposal_status == ProposalStatus::Replaced {
+            return None;
+        }
+        if let Some(previous) = &self.proposal {
+            if proposal != *previous {
+                // Mark status as replaced and extract equivocator (if known)
+                let equivocator = self.leader.as_ref().map(|leader| leader.key.clone());
+                warn!(
+                    ?equivocator,
+                    ?proposal,
+                    ?previous,
+                    "certificate proposal overrides local proposal (equivocation detected)"
+                );
+                self.proposal_status = ProposalStatus::Replaced;
+
+                // The votes we were tracking are not for this proposal, so we clear them.
+                self.notarizes.clear();
+                self.finalizes.clear();
+                self.proposal = Some(proposal);
+                return equivocator;
+            }
+        }
+
+        // Certificate proposals are considered verified (they have 2f+1 agreement)
+        if matches!(
+            self.proposal_status,
+            ProposalStatus::None | ProposalStatus::Unverified
+        ) {
+            debug!(?proposal, "setting verified proposal from certificate");
+            self.proposal_status = ProposalStatus::Verified;
         }
+
+        self.proposal = Some(proposal);
+        None
     }
 
-    async fn add_verified_notarize(&mut self, notarize: Notarize<S, D>) {
-        if self.proposal.is_none() {
+    async fn add_verified_notarize(&mut self, notarize: Notarize<S, D>) -> Option<S::PublicKey> {
+        // Check for equivocation
+        if self.proposal_status == ProposalStatus::Replaced {
+            return None;
+        }
+
+        // Check for conflicting proposal
+        if let Some(proposal) = self.proposal.as_ref() {
+            if proposal != &notarize.proposal {
```

Comment:
I think this should never happen because the batcher will lock-in to the first leader proposal it sees, so votes for a different proposal will never reach here IIRC. Either way still makes sense to have this logic.

Reply:
As discussed, this can happen if we get a certificate for some proposal before any notarize from the leader.